File: backend/agent/tools/misc.py
from __future__ import annotations

from typing import Any

from agent.tools.base import Tool


class SkillTool(Tool):
    name = "skill"
    description = "Load a local skill file and a small sample of nearby files."
    params = {"name": {"type": "string", "description": "Skill name"}}

    async def execute(self, args: dict[str, Any]) -> str:
        from agent.skills import find_skill, wrap_skill_content

        name = str(args.get("name", "")).strip()
        if not name:
            return "error: name is required"
        skill = find_skill(name)
        if skill is None:
            return f"error: skill not found: {name}"
        return wrap_skill_content(skill)


# ── todowrite ───────────────────────────────────────────────────────────────

_TODO_STORE: dict[str, list] = {}  # session_id → todos


class TodoWriteTool(Tool):
    name = "todowrite"
    description = (
        "Create and maintain a structured task list. "
        "Tracks progress and organizes multi-step work. "
        "States: pending, in_progress, completed, cancelled. "
        "Priorities: high, medium, low."
    )
    params = {
        "todos": {
            "type": "array",
            "description": "List of task items",
            "items": {
                "type": "object",
                "properties": {
                    "content": {"type": "string", "description": "Brief description of the task"},
                    "status": {
                        "type": "string",
                        "description": "pending | in_progress | completed | cancelled",
                    },
                    "priority": {"type": "string", "description": "high | medium | low"},
                },
                "required": ["content", "status", "priority"],
            },
        },
    }

    def __init__(self):
        super().__init__()
        self._session_id = "default"

    def set_session(self, session_id: str):
        self._session_id = session_id

    async def execute(self, args: dict[str, Any]) -> str:
        items = args.get("todos", [])
        if not items:
            return "error: todos list is empty"
        sid = self._session_id
        _TODO_STORE[sid] = items
        active = sum(1 for t in items if t.get("status") != "completed")
        return f"{active} todos — {len(items)} total:\n" + "\n".join(
            f"  [{t.get('status', '?')}] {t.get('priority', '?')}: {t.get('content', '')}"
            for t in items
        )


class QuestionTool(Tool):
    name = "question"
    description = (
        "Ask the user questions during a task. "
        "Use when you need clarification, preferences, or decisions. "
        "Supports multiple-choice and free-text questions. "
        "Each question can have options for the user to pick from."
    )
    params = {
        "questions": {
            "type": "array",
            "description": "Questions to ask the user",
            "items": {
                "type": "object",
                "properties": {
                    "question": {"type": "string", "description": "The question text"},
                    "header": {"type": "string", "description": "Short label (max 30 chars)"},
                    "options": {
                        "type": "array",
                        "description": "Available choices (omit for free-text answer)",
                        "items": {
                            "type": "object",
                            "properties": {
                                "label": {"type": "string", "description": "Display text"},
                                "description": {
                                    "type": "string",
                                    "description": "Explanation of choice",
                                },
                            },
                            "required": ["label"],
                        },
                    },
                    "multiple": {
                        "type": "boolean",
                        "description": "Allow selecting multiple choices",
                    },
                },
                "required": ["question", "header"],
            },
        },
    }

    async def execute(self, args: dict[str, Any]) -> str:
        raise RuntimeError("question tool must be handled by chat loop, not executed directly")

File: agent/tools/misc.go
package tools

import (
	"fmt"
	"strings"
)

// --- SkillTool ---

type SkillTool struct{}

func (t *SkillTool) Name() string { return "skill" }

func (t *SkillTool) Spec() ToolSpec {
	return ToolSpec{
		Description: "Load a skill by name.",
		Params: map[string]any{
			"name": prop("string", "Skill name to load"),
		},
		Required: []string{"name"},
	}
}

func (t *SkillTool) Execute(args map[string]any) Result {
	return R("error: skill tool not implemented in Go yet")
}

// --- TodoWriteTool ---

type TodoItem struct {
	Content  string `json:"content"`
	Status   string `json:"status"`
	Priority string `json:"priority"`
}

type TodoWriteTool struct {
	sessionID string
}

func NewTodoWriteTool() *TodoWriteTool {
	return &TodoWriteTool{sessionID: "default"}
}

func (t *TodoWriteTool) Name() string { return "todowrite" }

func (t *TodoWriteTool) Spec() ToolSpec {
	return ToolSpec{
		Description: "Update the session todo list.",
		Params: map[string]any{
			"todos": map[string]any{
				"type":        "array",
				"description": "Updated todo list",
				"items": map[string]any{
					"type": "object",
					"properties": map[string]any{
						"content":  prop("string", "Task description"),
						"status":   prop("string", "pending, in_progress, or completed"),
						"priority": prop("string", "high, medium, or low"),
					},
				},
			},
		},
		Required: []string{"todos"},
	}
}

func (t *TodoWriteTool) SetSession(id string) {
	t.sessionID = id
}

func (t *TodoWriteTool) Execute(args map[string]any) Result {
	raw, ok := args["todos"]
	if !ok {
		return R("error: todos list is required")
	}
	items, ok := raw.([]any)
	if !ok || len(items) == 0 {
		return R("error: todos list is empty")
	}
	var todos []TodoItem
	for _, r := range items {
		m, ok := r.(map[string]any)
		if !ok {
			continue
		}
		todos = append(todos, TodoItem{
			Content:  getStr(m, "content"),
			Status:   getStr(m, "status"),
			Priority: getStr(m, "priority"),
		})
	}
	if len(todos) == 0 {
		return R("error: todos list is empty")
	}
	active := 0
	for _, t := range todos {
		if t.Status != "completed" {
			active++
		}
	}
	var b strings.Builder
	b.WriteString(fmt.Sprintf("%d todos — %d total:\n", active, len(todos)))
	for _, t := range todos {
		b.WriteString(fmt.Sprintf("  [%s] %s: %s\n", t.Status, t.Priority, t.Content))
	}
	return R(strings.TrimRight(b.String(), "\n"))
}

// --- QuestionTool ---

type QuestionTool struct{}

func (t *QuestionTool) Name() string { return "question" }

func (t *QuestionTool) Spec() ToolSpec {
	return ToolSpec{
		Description: "Ask the user one or more structured questions and wait for answers.",
		Params: map[string]any{
			"questions": map[string]any{
				"type":        "array",
				"description": "Questions to ask the user",
				"items": map[string]any{
					"type": "object",
					"properties": map[string]any{
						"question": prop("string", "The question text"),
						"header":   prop("string", "Short label for the question"),
						"multiple": prop("boolean", "Allow selecting multiple options"),
						"options": map[string]any{
							"type":        "array",
							"description": "Optional answer choices",
							"items": map[string]any{
								"type": "object",
								"properties": map[string]any{
									"label":       prop("string", "Option label"),
									"description": prop("string", "Optional explanation of the option"),
								},
							},
						},
					},
				},
			},
		},
		Required: []string{"questions"},
	}
}

func (t *QuestionTool) Execute(args map[string]any) Result {
	return R("error: question tool needs interactive flow")
}
